### PR TITLE
FCCV Docs

### LINKED ISSUE #11538: enable ffcv with PL
A team introduced ffcv which does optimizations in the dataloader level.
As long as it can be a drop in replacement for dataloader it should add the benefits to any PyTorch and Lightning script.

ie:
```python
# current
data = Dataloader(...)

trainer.fit(..., data)

# 10x faster with ffcv
data = FFCVDataloader

trainer.fit(..., data)
```

Let's make sure when users want to get the benefits of both FFCV and PL they can do that!
cc @borda @rohitgr7 @Felonious-Spellfire @justusschock @awaelchli @ninginthecloud @otaj @tchaton @carmocca

### PR DESCRIPTION
**Pull request by justusschock**
## What does this PR do?

Adds docs for custom data iterables
Fixes #11538 

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

cc @borda @justusschock @awaelchli

### CODE REVIEW

**Review by carmocca — COMMENTED** [review]
If it's helpful, or just to share it, this is the script I used to work with FFCV (haven't run it in a while though):

<details>

```python
import os
from argparse import ArgumentParser
from contextlib import contextmanager
from time import perf_counter
from typing import Tuple, Optional

import torch
from pytorch_lightning import LightningModule, Trainer, LightningDataModule
from pytorch_lightning.utilities.imports import _module_available
from torch.utils.data import DataLoader, Dataset
from pytorch_lightning.utilities.types import TRAIN_DATALOADERS, EVAL_DATALOADERS


class RandomFFCVDataset(Dataset):
    def __init__(self, size, length):
        self.len = length
        self.data = torch.randn(length, size)

    def __getitem__(self, index):
        # convert to numpy: https://github.com/libffcv/ffcv/issues/101
        # return a tuple: https://github.com/libffcv/ffcv/issues/103
        return (self.data[index].numpy(),)

    def __len__(self):
        return self.len


class RandomDataset(Dataset):
    def __init__(self, size, length):
        self.len = length
        self.data = torch.randn(length, size)

    def __getitem__(self, index):
        return self.data[index]

    def __len__(self):
        return self.len


class BoringModel(LightningModule):
    def __init__(self, d):
        super().__init__()
        self.layer = torch.nn.Linear(d, 1)

    def forward(self, x):
        # return a tuple: https://github.com/libffcv/ffcv/issues/103
        if isinstance(x, tuple):
            x = x[0]
        return self.layer(x)

    def training_step(self, batch, batch_idx):
        loss = self(batch).sum()
        self.log("train_loss", loss)
        return {"loss": loss}

    def validation_step(self, batch, batch_idx):
        loss = self(batch).sum()
        self.log("valid_loss", loss)

    def test_step(self, batch, batch_idx):
        loss = self(batch).sum()
        self.log("test_loss", loss)

    # def on_train_epoch_end(self):
    #     print(torch.cuda.memory_summary(self.device))

    def configure_optimizers(self):
        return torch.optim.SGD(self.layer.parameters(), lr=0.1)


N, d = 2 ** 12, 128
batch_size = 8
num_workers = os.cpu_count()


_FFCV_AVAILABLE = _module_available("ffcv")
if _FFCV_AVAILABLE:
    from ffcv.loader import Loader, OrderOption
    from ffcv.writer import DatasetWriter
    from ffcv.fields import Field, NDArrayField
    import numpy as np


class FFCVMixin:
    def __init___(self):
        if not _FFCV_AVAILABLE:
            raise ModuleNotFoundError
        self.train_pipelines = ...

    @staticmethod
    def write_indexed_dataset(dataset: Dataset, path: str, fields: Mapping[str, Field]) -> DatasetWriter
        writer = DatasetWriter(path, fields)
        writer.from_indexed_dataset(dataset)


class BoringDataModule(LightningDataModule, FFCVMixin):
    def __init__(self) -> None:
        self.fit_dataset: Optional[Dataset] = None
        self.val_dataset: Optional[Dataset] = None
        self.test_dataset: Optional[Dataset] = None

    def setup(self, stage: Optional[str]=None) -> None:
        # https://docs.ffcv.io/writing_datasets.html#writing-a-dataset-to-ffcv-format
        fields = {"covariate": NDArrayField(shape=(d,), dtype=np.dtype("float32"))}
        if stage in (None, "fit"):
            self.fit_dataset = RandomFFCVDataset(d, N)
            self.write_indexed_dataset(self.fit_dataset, "fit_random", fields)
        if stage in (None, "val"):
            self.val_dataset = RandomFFCVDataset(d, N)
            self.write_indexed_dataset(self.val_dataset, "val_random", fields)
        if stage in (None, "test"):
            self.test_dataset = RandomFFCVDataset(d, N)
            self.write_indexed_dataset(self.test_dataset, "test_random", fields)

    def train_dataloader(self) -> TRAIN_DATALOADERS:
        return DataLoader(self.fit_dataset, batch_size=batch_size, num_workers=num_workers)

    def val_dataloader(self) -> EVAL_DATALOADERS:
        return DataLoader(self.val_dataset, batch_size=batch_size, num_workers=num_workers)

    def test_dataloader(self) -> EVAL_DATALOADERS:
        return DataLoader(self.test_dataset, batch_size=batch_size, num_workers=num_workers)


def prepare_data_ffcv(cwd: str, gpus: int) -> Tuple["Loader", "Loader", "Loader"]:
    # https://docs.ffcv.io/making_dataloaders.html#dataset-ordering
    if gpus == 0:
        # dataset fits in memory: fastest option
        kwargs = {"order": OrderOption.RANDOM}
    elif gpus == 1:
        # dataset does not fit in memory
        kwargs = {"order": OrderOption.QUASI_RANDOM, "os_cache": False}
    elif gpus > 1:
        # we assume dataset will fit in memory here
        # FIXME: distributed
        kwargs = {"order": OrderOption.RANDOM, "os_cache": True, "distributed": False}
    else:
        raise ValueError

    train_data = Loader(dataset_path, batch_size=batch_size, num_workers=num_workers, **kwargs)

    kwargs["order"] = OrderOption.SEQUENTIAL
    val_data = Loader(dataset_path, batch_size=batch_size, num_workers=num_workers, **kwargs)
    test_data = Loader(dataset_path, batch_size=batch_size, num_workers=num_workers, **kwargs)
    return train_data, val_data, test_data


@contextmanager
def catchtime() -> float:
    """https://stackoverflow.com/a/62956469"""
    start = perf_counter()
    yield lambda: perf_counter() - start


def run():
    cwd = os.getcwd()

    parser = ArgumentParser()
    parser.add_argument("--ffcv", action="store_true")
    parser.add_argument("--gpus", type=int, default=0)
    args = parser.parse_args()

    kwargs = {}
    if args.gpus:
        kwargs = {"accelerator": "gpu", "devices": args.gpus}
        if args.gpus > 1:
            kwargs["strategy"] = "ddp_find_unused_parameters_False"

    trainer = Trainer(
        default_root_dir=cwd,
        num_sanity_val_steps=0,
        max_epochs=10,
        enable_model_summary=False,
        enable_checkpointing=False,
        logger=False,
        enable_progress_bar=False,
        # callbacks=TQDMProgressBar(refresh_rate=100),
        benchmark=True,
        **kwargs,
    )

    model = BoringModel(d)
    datamodule = BoringDataModule()

    with catchtime() as t:
        trainer.fit(model, datamodule)
    if trainer.is_global_zero:
        print(f"Fit time: {t():.4f} secs")

    with catchtime() as t:
        trainer.test(model, datamodule, verbose=False)
    if trainer.is_global_zero:
        print(f"Test time: {t():.4f} secs")


if __name__ == "__main__":
    run()
```

</details>

File: docs/source-pytorch/data/custom_data_iterables.rst [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+    data = [(torch.rand(32, 3, 32, 32), torch.randint(0, 10, (32,))) for _ in range(100)]
+    model = LitClassifier()
+    trainer = Trainer()
+    trainer.fit(model, data)
+
+Examples for custom iterables include `NVIDIA DALI <https://github.com/NVIDIA/DALI>`__ or `FFCV <https://github.com/libffcv/ffcv>`__ for computer vision.
+Both libraries offer support for custom data loading and preprocessing (also hardware accelerated) and can be used with Lightning.
+
+
+For example taking the example from FFCV's readme, we can use it with Lightning by just replacing the hardcoded ``ToDevice(0)``
```
[the remark this answers is not in the document]
Comment (justusschock) [reply docs/source-pytorch/data/custom_data_iterables.rst:27]: I agree, but on the other hand, it doesn't really matter if this is done by us or ffcv, right? If they do it, ours is a no-op. I wanted to stick to their example as close as possible

File: docs/source-pytorch/data/custom_data_iterables.rst [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+
+.. code-block:: python
+
+    from ffcv.loader import Loader, OrderOption
+    from ffcv.transforms import ToTensor, ToDevice, ToTorchImage, Cutout
+    from ffcv.fields.decoders import IntDecoder, RandomResizedCropRGBImageDecoder
+
+
+    class CustomClf(LitClassifier):
+        def train_dataloader(self):
```
[the remark this answers is not in the document]
Comment (justusschock) [reply docs/source-pytorch/data/custom_data_iterables.rst:38]: I didn't mention it, since people are interested in how to integrate ffcv with PL not what requirements ffcv has and specifically as this is not related to training. Converting the dataset is a one-time thing and could be done separately.

File: docs/source-pytorch/data/custom_data_iterables.rst (hunk 1 above)
[the remark this answers is not in the document]
Comment (carmocca) [reply docs/source-pytorch/data/custom_data_iterables.rst:27]: I would remove it from this example that doesn't use it but mention how it would be used if needed. It's cool to show that it can be removed with PL

**Review by carmocca — APPROVED** [review]

**Review by Borda — APPROVED** [review]

**Review by awaelchli — APPROVED** [review]

**Review by akihironitta — APPROVED** [review]

File: docs/source-pytorch/data/custom_data_iterables.rst [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+``ToDevice(0)`` to ``ToDevice(self.trainer.local_rank)`` to correctly map to the desired GPU in your pipeline.
+
+.. code-block:: python
+
+    from ffcv.loader import Loader, OrderOption
+    from ffcv.transforms import ToTensor, ToDevice, ToTorchImage, Cutout
+    from ffcv.fields.decoders import IntDecoder, RandomResizedCropRGBImageDecoder
+
+
+    class CustomClf(LitClassifier):
```
Comment (lantiga) [inline docs/source-pytorch/data/custom_data_iterables.rst:38]:
```suggestion
    class CustomClassifier(LitClassifier):
```

File: docs/source-pytorch/data/custom_data_iterables.rst [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+                train_data = DALIGenericIterator(
+                    [get_dali_pipeline(batch_size=16)],
+                    ["data", "label"],
+                    reader_name="Reader",
+                )
+
+                return train_data
+
+
+Lightning works seamlessly with all kinds of custom data iterables,
```
Comment (lantiga) [inline docs/source-pytorch/data/custom_data_iterables.rst:106]: I don't think I fully understand this paragraph, can we make it more explicit and simpler to parse?

**Review by lantiga — CHANGES_REQUESTED** [review]

File: docs/source-pytorch/data/custom_data_iterables.rst (hunk 4 above)
Comment (justusschock) [reply docs/source-pytorch/data/custom_data_iterables.rst]: @lantiga I revisited this paragraph, is it better now?

File: docs/source-pytorch/data/custom_data_iterables.rst [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+
+- In a distributed multi-GPU setting (ddp),
+  Lightning automatically replaces the DataLoader's sampler with its distributed counterpart.
+  This makes sure that each GPU sees a different part of the dataset.
+  As sampling can be implemented in arbitrary ways with custom iterables,
+  there is no way for Lightning to know, how to replace the sampler.
+
+- When training fails for some reason, Lightning is able to extract all of the relevant data from the model,
+  optimizers, trainer and dataloader to resume it at the exact same batch it crashed.
+  This feature is called fault-tolerance and is limited to PyTorch DataLoaders as well as
```
Comment (lantiga) [inline docs/source-pytorch/data/custom_data_iterables.rst:120]:
```suggestion
  This feature is called fault-tolerance and is limited to PyTorch DataLoaders.
```

File: docs/source-pytorch/data/custom_data_iterables.rst [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+  Lighning also needs to know a lot about sampling, fast forwarding and random number handling to enable this,
```
Comment (lantiga) [inline docs/source-pytorch/data/custom_data_iterables.rst:121]:
```suggestion
  Lighning needs to know a lot about sampling, fast forwarding and random number handling to enable fault tolerance,
```

File: docs/source-pytorch/data/custom_data_iterables.rst [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+  meaning that this cannot be supported for arbitrary iterables either.
```
Comment (lantiga) [inline docs/source-pytorch/data/custom_data_iterables.rst:122]:
```suggestion
  meaning that it cannot be supported for arbitrary iterables.
```

**Review by lantiga — CHANGES_REQUESTED** [review]

**Review by lantiga — APPROVED** [review]

**Comment by lantiga** [comment]
Good to go!
